network/proto/blocks.go

```diff
@@ -0,0 +1,278 @@
+type trackingDAGBlocks struct {
+	blocks []*block
+}
+
+type muxDAGBlocks struct {
+	Underlying DAGBlocks
+	mux        *sync.Mutex
+}
+
+func (m muxDAGBlocks) String() string {
+	m.mux.Lock()
+	defer m.mux.Unlock()
+	return m.Underlying.String()
+}
+
+func (m muxDAGBlocks) Get() []DAGBlock {
+	m.mux.Lock()
+	defer m.mux.Unlock()
+	return m.Underlying.Get()
+}
+
+func (m muxDAGBlocks) AddTransaction(tx dag.SubscriberTransaction, payload []byte) error {
+	m.mux.Lock()
+	defer m.mux.Unlock()
+	return m.Underlying.AddTransaction(tx, payload)
+}
+
+type block struct {
+	start time.Time
+	// key: tx ref, value: distance (in blocks) of the next TX
+	heads map[hash.SHA256Hash]*head
+}
+
+type head struct {
+	distance    int
+	signingTime time.Time
+	// blockDate contains the signing time at start of the block
+	blockDate time.Time
+}
+
+// heads calculates the block heads, without updating the structure first. Only for internal use and testting.
+func (blx *trackingDAGBlocks) heads() []DAGBlock {
+	result := make([]DAGBlock, 0)
+	for blockNum, currBlock := range blx.blocks {
+		resultBlock := DAGBlock{Start: currBlock.start}
+		for ref := range currBlock.heads {
+			if blockNum < len(blx.blocks) {
+				resultBlock.Heads = append(resultBlock.Heads, ref)
+			}
+		}
+		result = append(result, resultBlock)
+	}
+	return result
+}
+
+// Heads returns the block heads.
+func (blx *trackingDAGBlocks) Get() []DAGBlock {
+	blx.update(time.Now())
+	return blx.heads()
+}
+
+// AddTransaction adds a transaction to the DAG blocks structure. It MUST with the transactions in order, so it's
+// typically called using a sequential DAG subscriber. So given TXs `A <- B <- [C, D]` call order is A, B, C, D (or A, B, D, C).
+func (blx *trackingDAGBlocks) AddTransaction(tx dag.SubscriberTransaction, _ []byte) error {
+	blx.update(time.Now())
+	// Determine block the TX is part of
+	blockIdx := len(blx.blocks) - 1
+	for i := 0; i < len(blx.blocks)-1; i++ {
+		// Lower blocks are earlier, so when the TXs time is before the start time of the next block, this is the block
+		// the TX is part of.
+		if tx.SigningTime().Before(blx.blocks[i+1].start) {
+			blockIdx = i
+			break
+		}
+	}
+	txBlock := blx.blocks[blockIdx]
+	// Prevs of this TX in this block were previously heads (prev'd by another branch within this block) but now
+	// the current TX will be the new head. So we 'un-head' all prevs and mark the current TX as head. This works as long
+	// as this func is called with TXs in order.
+	txBlockDate := startOfDay(tx.SigningTime())
+	for _, prev := range tx.Previous() {
+		if prevTX, ok := txBlock.heads[prev]; ok {
+			// But not if this the signing time of the tx would put it in a future block, which can be the case
+			// when tx's with future timestamps are added in today's block
+			if !txBlockDate.After(prevTX.blockDate) {
+				delete(txBlock.heads, prev)
+			}
+		}
+	}
+	txBlock.heads[tx.Ref()] = &head{
+		distance:    math.MaxInt64,
```

Comment:
when shifting TXs into an older block it decreases the `distance` property. If a TX is added which `prev`s the TX and has a shorter distance, it takes the shortest distance.
```
if newDistance < head.distance {
   head.distance = newDistance
}
```

With `MaxInt64` as default value TXs without a next TX are never removed as head (since `MaxInt64` means a TX has a signing time thousands of millenia in the future), and any newly added TX that has it as `prev` leads to a shorter `distance`.

Could've taken `-1` but this is leads to less code and no bounds checking.